File: swig/matlab/pireceiver.m
% PIRECEIVER script demonstrating MATLAB-HELICS interface
%
% See pisender for usage information
%
% This example attempts to request the next time equal to its
% sim_stop_time, but the corresponding pisender will request intermediate
% times such that the pireceiver will be granted times earlier than
% requested
%
% Note to non-programmers: if unfamiliar, the assert functions simply check
% if the condition is true and if so continues. If false, execution stops.

%% Initialize HELICS library in MATLAB
helicsStartup()

%% Configuration
deltat = 0.01;  %Base time interval (seconds)
sim_stop_time = 20;

% HELICS options
helics_core_type = 'zmq'; 
fedinitstring = '--federates=1';  % required with current C interface when using separate processes for each federate

%% Provide summary information
helicsversion = helics.helicsGetVersion();

fprintf('PI RECEIVER: Helics version = %s\n', helicsversion)

%% Create Federate Info object that describes the federate properties
fedinfo = helics.helicsFederateInfoCreate();
assert(not(isempty(fedinfo)))

% Set Federate name
status = helics.helicsFederateInfoSetFederateName(fedinfo, 'MATLAB Pi Receiver Federate');
assert(status==0)

% Set core type from string
status = helics.helicsFederateInfoSetCoreTypeFromString(fedinfo, helics_core_type);
assert(status==0)

% Federate init string
status = helics.helicsFederateInfoSetCoreInitString(fedinfo, fedinitstring);
assert(status==0)

%% Set the message interval (timedelta) for federate. 
% Note:
% HELICS minimum message time interval is 1 ns and by default
% it uses a time delta of 1 second. What is provided to the
% setTimedelta routine is a multiplier for the default timedelta 
% (default unit = seconds).

% Set one message interval
status = helics.helicsFederateInfoSetTimeDelta(fedinfo, deltat);
assert(status==0)

status = helics.helicsFederateInfoSetLoggingLevel(fedinfo, 1);
assert(status==0)

%% Actually create value federate
vfed = helics.helicsCreateValueFederate(fedinfo);
disp('PI RECEIVER: Value federate created');

% Subscribe to PI SENDER's publication (note: published as global)
sub = helics.helicsFederateRegisterSubscription(vfed, 'testA', 'double', '');

disp('PI RECEIVER: Subscription registered (testA)');

%% Start execution
status = helics.helicsFederateEnterExecutionMode(vfed);
if status == 0
    disp('PI RECEIVER: Entering execution mode');
else
    error('PI RECEIVER: Failed to enter execution mode (status = %d)\n Try running pisender.m first. (or start the broker seperately)', status);
end

%% Execution Loop
granted_time =- 1;  %Force at least one run

%% Continue execution until end of requested simulation time
while granted_time <= sim_stop_time

    [status, granted_time] = helics.helicsFederateRequestTime(vfed, sim_stop_time);
    assert(status==0)

    isupdated = helics.helicsSubscriptionIsUpdated(sub);

    if (isupdated == 1)
        [result, value] = helics.helicsSubscriptionGetDouble(sub);
        fprintf('PI RECEIVER: Received value = %g at time %4.1f from PI SENDER\n', value, granted_time);
    end
end

%% Shutdown
helicsDestroyFederate(vfed);

disp('PI RECEIVER: Federate finalized');

File: swig/python/pireceiver.py
import helics as h

fedinitstring = "--broker=mainbroker --federates=2"
deltat = 0.01

helicsversion = h.helicsGetVersion()

print("PI RECEIVER: Helics version = {}".format(helicsversion))

# Create Federate Info object that describes the federate properties */
print("PI RECEIVER: Creating Federate Info")
fedinfo = h.helicsFederateInfoCreate()

# Set Federate name
print("PI RECEIVER: Setting Federate Info Name")
status = h.helicsFederateInfoSetFederateName(fedinfo, "TestB Federate")

# Set core type from string
print("PI RECEIVER: Setting Federate Info Core Type")
status = h.helicsFederateInfoSetCoreTypeFromString(fedinfo, "zmq")

# Federate init string
print("PI RECEIVER: Setting Federate Info Init String")
status = h.helicsFederateInfoSetCoreInitString(fedinfo, fedinitstring)

# Set the message interval (timedelta) for federate. Note that
# HELICS minimum message time interval is 1 ns and by default
# it uses a time delta of 1 second. What is provided to the
# setTimedelta routine is a multiplier for the default timedelta.

# Set one second message interval
print("PI RECEIVER: Setting Federate Info Time Delta")
status = h.helicsFederateInfoSetTimeDelta(fedinfo, deltat)

print("PI RECEIVER: Setting Federate Info Logging")
status = h.helicsFederateInfoSetLoggingLevel(fedinfo, 1)

# Create value federate
print("PI RECEIVER: Creating Value Federate")
vfed = h.helicsCreateValueFederate(fedinfo)
print("PI RECEIVER: Value federate created")

# Subscribe to PI SENDER's publication
sub = h.helicsFederateRegisterSubscription(vfed, "testA", "double", "")
print("PI RECEIVER: Subscription registered")

status = h.helicsFederateEnterExecutionMode(vfed)
print("PI RECEIVER: Entering execution mode")

value = 0.0
prevtime = 0

currenttime = h.helicsFederateRequestTime(vfed, 0.19)

while (currenttime <= 0.19):

    currenttime = h.helicsFederateRequestTime(vfed, 0.19)

    isupdated = h.helicsSubscriptionIsUpdated(sub)

    if (isupdated == 1) and (currenttime > prevtime):
        result, value = h.helicsGetDouble(sub)
        print("PI RECEIVER: Received value = {} at time {} from PI SENDER".format(value, currenttime))

    prevtime = currenttime

status = h.helicsFederateFinalize(vfed)

h.helicsFederateFree(vfed)
h.helicsCloseLibrary()
print("PI RECEIVER: Federate finalized")
